scientific_library/tvb/simulator/simulator.py

```diff
@@ -387,20 +419,50 @@ def __call__(self, simulation_length=None, random_state=None):
         stimulus = self._prepare_stimulus()
         state = self.current_state
 
+        # Do for initial condition:
+        step = self.current_step + 1  # the first step in the loop
+        node_coupling = self._loop_compute_node_coupling(step)
+        self._loop_update_stimulus(step, stimulus)
+        # This is not necessary in most cases
+        # if update_non_state_variables=True in the model dfun by default
+        self.update_state(state, node_coupling, local_coupling)
+
         # integration loop
         n_steps = int(math.ceil(self.simulation_length / self.integrator.dt))
+        tic = time.time()
+        tic_ratio = 0.1
+        tic_point = tic_ratio * n_steps
         for step in range(self.current_step + 1, self.current_step + n_steps + 1):
-            # needs implementing by hsitory + coupling?
+            # Integrate TVB to get the new TVB state
+            state = self.integrator.scheme(state, self.model.dfun, node_coupling, local_coupling, stimulus)
+            # Prepare coupling and stimulus for next time step
+            # and, therefore, for the new TVB state:
             node_coupling = self._loop_compute_node_coupling(step)
             self._loop_update_stimulus(step, stimulus)
-            state = self.integrator.scheme(state, self.model.dfun, node_coupling, local_coupling, stimulus)
+            # Update any non-state variables and apply any boundaries again to the new state:
+            self.update_state(state, node_coupling, local_coupling)
+            # Now direct the new state to history buffer and monitors
             self._loop_update_history(step, n_reg, state)
             output = self._loop_monitor_output(step, state)
             if output is not None:
                 yield output
+            if step - self.current_step >= tic_point:
+                toc = time.time() - tic
+                if toc > 600:
+                    if toc > 7200:
+                        time_string = "%0.1f hours" % (toc / 3600)
+                    else:
+                        time_string = "%0.1f min" % (toc / 60)
+                else:
+                    time_string = "%0.1f sec" % toc
+                print_this = "\r...%0.1f%% done in %s" % \
+                             (100.0 * (step - self.current_step) / n_steps, time_string)
+                sys.stdout.write(print_this)
+                sys.stdout.flush()
```

Comment:
This would be good to integrate into the [ProgressLogger](https://github.com/the-virtual-brain/tvb-root/blob/master/scientific_library/tvb/simulator/monitors.py#L911) monitor maybe taking cues from [tqdm](https://github.com/tqdm/tqdm)? but not here

Reply:
I just put it there so that you see it and move it somewhere else. It is something lacking currently.